### PR TITLE
OpenAPI conversion validation as a PR gate

### LINKED ISSUE #62: XSLT rules can break Edm or OpenAPI model parsing
Our checkins for XSLT rules are not blocked by a PR gate that checks whether our transformations are still producing parsable outputs.

- Typewriter uses EDM model parsing
- Powershell uses OpenAPI conversion 
[AB#7832](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/7832)

### PR DESCRIPTION
**Pull request by zengin**
- adds OpenAPI reader as a submodule
- adds a step to MetadataParser tool to convert to OpenAPI
- adds a Github Action which will
  - apply the latest XSLT rules on the latest v1 and beta metadata
  - validate those transformed metadata is parsable

Fixes #62

### CODE REVIEW

**Comment by zengin** [comment]
Tested the new validation rule with the broken XSLT and it failed. Reverted back that change and it succeeded.

**Comment by zengin** [comment]
@baywet can you have a look at the Github Action failure for updating latest OpenAPI descriptions? Not sure why the push fails.

File: tools/MetadataParser/Program.cs [hunk 1]
```diff
@@ -1,12 +1,35 @@
 ﻿using Microsoft.OData.Edm.Csdl;
 using System.Xml;
 using System;
+using Microsoft.OData.Edm;
```
Comment (MIchaelMainer) [inline tools/MetadataParser/Program.cs:4]: nit: file should contain copyright header

**Review by MIchaelMainer — APPROVED** [review]

**Review by nikithauc — APPROVED** [review]

File: scripts/run-metadata-validation.ps1 [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+.Parameter repoDirectory
+    Full path the the root directory of msgraph-metadata checkout.
+#>
+
+param(
+    [Parameter(Mandatory=$true)][string]$repoDirectory
+)
+
+$transformCsdlDirectory = Join-Path $repoDirectory "transforms/csdl"
```
Comment (baywet) [inline scripts/run-metadata-validation.ps1:26]:
```suggestion
$transformCsdlDirectory = Join-Path $repoDirectory "transforms" "csdl"
```

File: scripts/run-metadata-validation.ps1 [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+$transformScript = Join-Path $transformCsdlDirectory "transform.ps1"
+$xsltPath = Join-Path $transformCsdlDirectory "preprocess_csdl.xsl"
+
+$betaSnapshot = Join-Path $repoDirectory "beta_metadata.xml"
+$v1Snapshot = Join-Path $repoDirectory "v1.0_metadata.xml"
+
+$metadataParserTool = Join-Path $repoDirectory "tools/MetadataParser/MetadataParser.csproj"
```
Comment (baywet) [inline scripts/run-metadata-validation.ps1:33]:
```suggestion
$metadataParserTool = Join-Path $repoDirectory "tools" "MetadataParser" "MetadataParser.csproj"
```

File: scripts/run-metadata-validation.ps1 [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+$transformedBeta = Join-Path $repoDirectory "transformed_beta_metadata.xml"
+$transformedV1 = Join-Path $repoDirectory "transformed_v1.0_metadata.xml"
+
+Write-Host "Tranforming beta metadata using xslt..."
+& $transformScript -xslPath $xsltPath -inputPath $betaSnapshot -outputPath $transformedBeta
+
+Write-Host "Validating beta metadata after the transform..."
+& dotnet run -p $metadataParserTool $transformedBeta
```
Comment (baywet) [inline scripts/run-metadata-validation.ps1:42]: nit: I'd move the dotnet run steps in the workflow, this way we'd have a better integration for linting, warning etc... with the UI

**Review by baywet — CHANGES_REQUESTED** [review]

**Comment by baywet** [comment]
@zengin fixed the pipeline (I used your PR as guinea pig)!

**Review by peombwa — APPROVED** [review]

File: scripts/run-metadata-validation.ps1 (hunk 2 above)
Comment (zengin) [reply scripts/run-metadata-validation.ps1:26]: Unfortunately `Join-Path` doesn't work with varargs and you need [weird tricks in piping](http://codebuckets.com/2016/08/20/powershell-join-path-with-multiple-parameters/) or multiple levels of function composition to get it working in single line.

Forward slash works cross platform.

File: scripts/run-metadata-validation.ps1 (hunk 3 above)
Comment (zengin) [reply scripts/run-metadata-validation.ps1:33]: See other comment

File: scripts/run-metadata-validation.ps1 (hunk 4 above)
Comment (zengin) [reply scripts/run-metadata-validation.ps1:42]: I would like to be able to run this script as a self contained test in local dev box without needing to create a PR and then realizing that my XSLT is wrong. That is the reason I am keeping everything in a single ps1 file. I wish Actions had a way to run as is on a local dev environment without much modifications.

**Comment by zengin** [comment]
@baywet I added a quick fix to allow empty commits in OpenAPI action. Thanks for the other fixes you pushed.

I have addressed your comments. Can you please have another look?

File: scripts/run-metadata-validation.ps1 (hunk 4 above)
Comment (baywet) [reply scripts/run-metadata-validation.ps1:42]: yeah not having the ability to run workflows locally is a real pain. I pushed 18 commits to fix the workflow before squashing and force push....

you could have two scripts:
- one that calculates the paths and settings them as tasks output (this one
- one that calls the first one and dotnet run for the local debug experience

Comment (zengin) [reply scripts/run-metadata-validation.ps1:42]: I think this is a script that is short enough to be single file. And local testing instruction is as simple as:
> After you made your changes to the XSLT, run:
> `./scripts/run-metadata-validation.ps1 <local_metadata_checkout_folder>`

And everything is maintained in a single place.

**Review by baywet — APPROVED** [review]

**Pull request merged**